File: product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output = []
        pre = []
        post = [0] * len(nums)
        mult = 1

        # Product of numbers before the indice
        for i in range(0, len(nums)):
            if i == 0:
                pre.append(1)
            else:
                mult *= nums[i-1]
                pre.append(mult)

        # Product of numbers after the indice
        mult = 1
        for i in range(len(nums)-1, -1, -1):
            if i == len(nums)-1:
                post[i] = 1
            else:
                mult *= nums[i+1]
                post[i] = mult
        
        # Product of both before and after indices
        for i in range(len(nums)):
            output.append(pre[i]*post[i])
        return output

File: product-of-array-except-self/product-of-array-except-self.java
class Solution 
{
    public int[] productExceptSelf(int[] nums) 
    {
        int[] answer = new int[nums.length];
        answer[0] = 1;
        
        // Accumulate products from left -> right
        for (int i = 1; i < nums.length; i++)
            answer[i] = answer[i-1] * nums[i-1];
        
        // Accumulate other products from right -> left
        int back = 1;
        for (int i = nums.length-1; i > -1; i--)
        {
            answer[i] *= back;
            back *= nums[i];
        }
        
        return answer;
    }
}